File: findDuplicate.cpp
#include <bits/stdc++.h>
using namespace std;

class SolutionBrute
{
public:
    int findDuplicate(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 1; i++)
        {
            if (nums[i] == nums[i + 1])
            {
                return nums[i];
            }
        }
        return -1; // Return a default value if no duplicate is found
    }
};

class SolutionBetter
{
public:
    int findDuplicate(vector<int> &nums)
    {
        int n = nums.size();
        vector<int> freq(n + 1, 0);
        for (int i = 0; i < n; i++)
        {
            if (freq[nums[i]] == 0)
            {
                freq[nums[i]]++;
            }
            else
            {
                return nums[i];
            }
        }
        return -1; // Return a default value if no duplicate is found
    }
};

class SolutionOptimal
{
    public:
        int findDuplicate(vector<int> &nums)
    {
        int n = nums.size();
        int slow = nums[0];
        int fast = nums[0];

        
    }
};

int main()
{
    SolutionBrute sol;
    vector<int> nums = {3, 1, 3, 4, 2};
    int duplicate = sol.findDuplicate(nums);
    cout << "Duplicate number is: " << duplicate << endl;

    SolutionBetter solBetter;
    nums = {3, 1, 3, 4, 2};
    duplicate = solBetter.findDuplicate(nums);
    cout << "Duplicate number is: " << duplicate << endl;

    return 0;
}

File: Linked List/SLL/find_duplicate.py
from typing import List

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow = nums[0]
        fast = nums[0]
        
        # Phase 1: Finding the intersection point of the two runners.
        
        while True:
            slow = nums[slow]  # Move slow pointer by 1 step
            fast = nums[nums[fast]]  # Move fast pointer by 2 steps
            if slow == fast:  # A cycle is detected
                break
            
        # Phase 2: Finding the entrance to the cycle.
        fast = nums[0]  # Reset fast pointer to the start of the array or slow pointer to the start of the array

        while slow != fast:
            slow = nums[slow]  # Move slow pointer by 1 step
            fast = nums[fast]  # Move fast pointer by 1 step
            
        return slow  # The duplicate number is found
    
    
# Example usage
solution = Solution()

# Test case 1
nums1 = [1, 3, 4, 2, 2]
print(solution.findDuplicate(nums1))  # Output: 2